# Entrega 2 - Árboles de Decisión

### Grupo M:
     - Leonel Rosano  5.039.791-0
     - Emiliano Pérez 4.787.149-2
     - Felipe Chavat 4.659.472-2



## 1. Objetivo

El objetivo de este laboratorio es construir Árboles de Decisión utilizando el algorítmo _ID3_ y modificandolo para que sea capaz de soportar atributos numéricos. En primera instancia se construye un Árbol de Decisión capaz de predecir clases del conjunto _Iris_. Luego se genera un arbol por cada clase, donde cada árbol tomara el criterio de clasificacion "una clase versus el resto".
Finalmente se repite el procedimiento haciendo uso del conjunto _Covertype_.

## 2. Primera parte

Para que _ID3_ soporte atributos con valores numéricos continuos fue necesario modificar el algorítmo para que sea capaz de discretizar un atributo con estas especificaciones.
Para ello, a la hora de considerar un atributo cuantitativo se utilizó un umbral para subdividir el conjunto de entrenamiento en dos (por ejemplo, mayores y menores o igual)
Este umbral se obtiene de un conjunto de posibles umbrales calculados de la siguiente forma:
Dado un atributo (A), el cual se quiere saber si es la mejor opción en la iteración actual, se ordena el conjunto de entrenamiento (S) según el atributo A y se toma el valor medio de este atributo entre dos tuplas consiguientes cuya clase es distinta.
De estos posibles umbrales, el mejor es aquel que maximiza la ganancia.

Para medir la ganancia de información de un atributo, el algorítmo _ID3_ utiliza la entropía (_H(S)_), y se calcula la ganancia como la diferencia entre la entropía del conjunto actual y la entropía del conjunto resultante de subdividirlo según el atributo. Sin embargo, vimos que es posible calcular la ganancia de información como la diferencia de la _impureza de Gini_[PONER REFERENCIA] ó la diferencia 

### 2.1 Preparación del conjunto de entrenamiento

El conjunto _Iris_ consta de 150 instancias en total, las cuales clasifican a la planta Iris en 3 distintas especies: Iris Setosa, Iris Versicolour, Iris Virginica. Los atributos utilizados para clasificar la especie son 4, todos numéricos continuos, por ende, será necesario utilizar el algorítmo ID3 modificado para lograr generar un árbol de decisión capaz de clasificarlas.

Dado el conjunto de datos _Iris_, se reorganizan las tuplas de forma aleatoria para lograr una buena distribución de las clases a clasificar. Luego se particiona este conjunto en dos subconjuntos A y B correspondientes al 60% y 40% del conjunto original.

El subconjunto A es utilizado para entrenar el Árbol de Decisión y el subconjunto B se utiliza para hacer una evaluación del modelo resultante. No se evalúa el modelo con el mismo conjunto con el que se entrena ya que, de ser así, podríamos estar dentro de un caso de sobre ajuste y no seríamos capaces de identificarlo.


### 2.2 Evaluación de los modelos realizados

Se generan tres modelos distintos usando para cada uno de estos una métrica distinta a la hora de hallar la ganancia de información para cada atributo.

Los resultados que se muestran tienen el siguiente formato:
En primera instancia, se mostrará el arbol con toda la información obtenida en cada nodo. Por ejemplo, el nodo
`=> 2 >= 3.3, Gain: 0.898 , MCV:  1 , P:  34.286 % --- COE:  [(1, 36), (2, 36), (3, 33)]`
indica que es no es una hoja (_=>_), si lo fuera estaría representado por '_->_', la condición para subdividir el conjunto (atributo _2 >= 3.3_), la ganancia generada por el atributo (_Gain_), la clase mas comunmente observada en el conjunto de entrenamiento actual (_MCV_), el porcentaje de la cantidad de resultados con el valor mas comunmente observado (P) y la cantidad de ocurrencias de cada clase en el conjunto de entrenamiento actual (_COE_).

Luego, se podrá observar una representación visual del árbol generado para tener una mejor idea de la estructura de este.

Por último, se presentará una matriz de confusión para poder hacer observación de que tan bueno es el modelo obtenido.

#### 2.2.1 Usando Entropía

En primera instancia, se genera un árbol utilizando entropía (_Shannon Entropy_) para calcular la ganancia. Los resultados obtenidos son los siguientes:

In [5]:
from main import makeSimpleTree
tree = makeSimpleTree('entropy')

=> 2 >= 2.6, Gain: 0.898 , MCV:  1 , P:  34.286 % --- COE:  [(1, 36), (2, 36), (3, 33)]
 -> Class: 3 , MCV:  3 , P:  100.0 % --- COE:  [(3, 33)]
 => 3 >= 1.7, Gain: 0.753 , MCV:  1 , P:  50.0 % --- COE:  [(1, 36), (2, 36)]
  => 2 >= 5.3, Gain: 0.125 , MCV:  1 , P:  94.595 % --- COE:  [(1, 35), (2, 2)]
   => 0 >= 4.9, Gain: 0.128 , MCV:  1 , P:  97.222 % --- COE:  [(1, 35), (2, 1)]
    => 1 >= 2.45, Gain: 1.0 , MCV:  1 , P:  50.0 % --- COE:  [(1, 1), (2, 1)]
     -> Class: 1 , MCV:  1 , P:  100.0 % --- COE:  [(1, 1)]
     -> Class: 2 , MCV:  2 , P:  100.0 % --- COE:  [(2, 1)]
    -> Class: 1 , MCV:  1 , P:  100.0 % --- COE:  [(1, 34)]
   -> Class: 2 , MCV:  2 , P:  100.0 % --- COE:  [(2, 1)]
  => 2 >= 4.8, Gain: 0.108 , MCV:  2 , P:  97.143 % --- COE:  [(1, 1), (2, 34)]
   => 0 >= 5.95, Gain: 0.918 , MCV:  2 , P:  66.667 % --- COE:  [(1, 1), (2, 2)]
    -> Class: 1 , MCV:  1 , P:  100.0 % --- COE:  [(1, 1)]
    -> Class: 2 , MCV:  2 , P:  100.0 % --- COE:  [(2, 2)]
   -> Class: 2 , MCV:

La matriz de confusión obtenida para este caso es:

|-|Iris Setosa |Iris Versicolour |Iris Virginica |
|---:|---:|---:|---:|
| Iris Setosa |**8** |0 |0 |
| Iris Versicolour |2 |**3** |0 |
| Iris Virginica |0 |0 |**12** |

Valores macro y micro

|- |Prec |Rec |Fs 0.5 |
|---: |---: |---: |---:|
|Micro |0.92 |0.92 |0.92|
|Macro |0.933 |0.867 |0.899|

#### 2.2.2 Usando impureza de Gini

La impureza de Gini nos da noción de que tan probable es que dos elementos tomados aleatoriamente sean distintos. Cuanto menor sea el valor resultante, mas 'puro' es el conjunto el cual estamos midiendo su impureza.

Por ejemplo, dado un conjunto con dos clases distintas

| #Class1 | #Class2 | Gini |
|---:|---:|------:|
|0|5|0|
|1|1|0.5|
|1|2|0.44|


Los resultados obtenidos haciendo uso de la impureza de Gini son:

In [2]:
tree = makeSimpleTree('gini')

=> 2 >= 2.6, Gain: 0.323 , MCV:  1 , P:  34.286 % --- COE:  [(1, 36), (2, 36), (3, 33)]
 -> Class: 3 , MCV:  3 , P:  100.0 % --- COE:  [(3, 33)]
 => 3 >= 1.7, Gain: 0.42 , MCV:  1 , P:  50.0 % --- COE:  [(1, 36), (2, 36)]
  => 2 >= 5.3, Gain: 0.05 , MCV:  1 , P:  94.595 % --- COE:  [(1, 35), (2, 2)]
   => 0 >= 4.9, Gain: 0.026 , MCV:  1 , P:  97.222 % --- COE:  [(1, 35), (2, 1)]
    => 1 >= 2.45, Gain: 0.5 , MCV:  1 , P:  50.0 % --- COE:  [(1, 1), (2, 1)]
     -> Class: 1 , MCV:  1 , P:  100.0 % --- COE:  [(1, 1)]
     -> Class: 2 , MCV:  2 , P:  100.0 % --- COE:  [(2, 1)]
    -> Class: 1 , MCV:  1 , P:  100.0 % --- COE:  [(1, 34)]
   -> Class: 2 , MCV:  2 , P:  100.0 % --- COE:  [(2, 1)]
  => 2 >= 4.8, Gain: 0.017 , MCV:  2 , P:  97.143 % --- COE:  [(1, 1), (2, 34)]
   => 0 >= 5.95, Gain: 0.444 , MCV:  2 , P:  66.667 % --- COE:  [(1, 1), (2, 2)]
    -> Class: 1 , MCV:  1 , P:  100.0 % --- COE:  [(1, 1)]
    -> Class: 2 , MCV:  2 , P:  100.0 % --- COE:  [(2, 2)]
   -> Class: 2 , MCV:  

La matriz de confusión obtenida para este caso es:

|-|Iris Setosa |Iris Versicolour |Iris Virginica |
|---:|---:|---:|---:|
| Iris Setosa |**8** |0 |0 |
| Iris Versicolour |2 |**3** |0 |
| Iris Virginica |0 |0 |**12** |

Valores macro y micro

|- |Prec |Rec |Fs 0.5|
|---: |---: |---: |---:|
|Micro |0.92 |0.92 |0.92|
|Macro |0.933 |0.867 |0.899|

#### 2.2.3 Usando misclassification

In [3]:
tree = makeSimpleTree('misclassification')

=> 2 >= 2.6, Gain: 0.314 , MCV:  1 , P:  34.286 % --- COE:  [(1, 36), (2, 36), (3, 33)]
 -> Class: 3 , MCV:  3 , P:  100.0 % --- COE:  [(3, 33)]
 => 3 >= 1.7, Gain: 0.458 , MCV:  1 , P:  50.0 % --- COE:  [(1, 36), (2, 36)]
  => 2 >= 5.3, Gain: 0.027 , MCV:  1 , P:  94.595 % --- COE:  [(1, 35), (2, 2)]
   -> Class: 1 , MCV:  1 , P:  97.222 % --- COE:  [(1, 35), (2, 1)]
   -> Class: 2 , MCV:  2 , P:  100.0 % --- COE:  [(2, 1)]
  -> Class: 2 , MCV:  2 , P:  97.143 % --- COE:  [(1, 1), (2, 34)]


    ______2
   /       \
  3__       3
 /   \
2     2
     / \
    2   1



Matriz de confusión resultante:


|-|Iris Setosa |Iris Versicolour |Iris Virginica |
|---:|---:|---:|---:|
| Iris Setosa |**8** |0 |0 |
| Iris Versicolour |2 |**3** |0 |
| Iris Virginica |0 |0 |**12** |

Valores macro y micro

|- |Prec |Rec |Fs 0.5|
|---: |---: |---: |---:|
|Micro |0.92 |0.92 |0.92|
|Macro |0.933 |0.867 |0.899|

#### 2.2.4 5-fold cross validation

Se realiza 5-fold cross validation y se evalúa cada modelo obtenido hallando la media armónica de cada uno de estos.
Es importante aclarar que para hallar la medida armónica se utiliza el conjunto destinado a evaluación del modelo generado.
Los resultados obtenidos son:

| - | Micro F Score | Macro F Score |
|--:|------------:|------------:|
| 0 | **1.0** | **1.0** |
| 1 | **1.0** | **1.0** |
| 2 | **1.0** | **1.0** |
| 3 | **1.0** | **1.0** |
| 4 | 0.92 | 0.8987654320987655 |

Pudimos observar que en la mayoría de los modelos generados se obtiene el mejor resultado posible. Creemos que se debe a que en los conjuntos de entrenamiento generados el porcentaje de casos distintos fue consistente además de que el algorítmo implementado es suficientemente robusto para saber clasificar estos casos. 

Al final, utilizamos un conjunto con un total de 20 tuplas que no fue utilizado ni para entrenar ni para evaluar, con el objetivo de asegurar que no estamos en casos de sobre ajuste. Los resultados son los siguientes:

| - | Micro Score | Macro Score |
|--:|------------:|------------:|
| 0 | **1.0** | **1.0** |
| 1 | **1.0** | **1.0** |
| 2 | **1.0** | **1.0** |
| 3 | **1.0** | **1.0** |
| 4 | 0.9 | 0.901565995525727 |


### 2.3 Conclusiones

Fuimos capaces de clasificar de forma correcta el 92% de las tuplas del conjunto de evaluación, por lo que consideramos que los modelos realizados tienen un buen resultado.
También creemos que la cantidad de tuplas que contiene el conjunto utilizado (conjunto _Iris_) no es tan grande como para poder reconocer posibles diferencias a la hora de utilizar distintas métricas. Por la misma razón, consideramos no realizar post procesamiento (post-pruning) a los árboles obtenidos.

## 3. Segunda parte

Para la realización de este paso, se implementa un algorítmo (clase llamada _PoolTree_) capaz de almacenar un modelo de uno versus todos para cada clase existente en el conjunto de datos, y que es capaz de ejecutar una clasificación en función de estos modelos.
Para el caso del conjunto _Iris_, el resultado es una instancia de la clase _PoolTree_ que contiene tres modelos distintos, uno que clasifica a las _Iris Setosa_ versus las demas, y analogamente los otros dos modelos con _Iris Versicolour_ e _Iris Virginica_.
A la hora de implementar la clasificación de este conjunto de modelos, fue necesario definir un criterio para toma de decisiones en caso de que existiese un conflicto de predicción (mas de un modelo reclamando la tupla a predecir).

### 3.1 Criterio de Clasificación

A cada nodo del arbol generado se le asignó el porcentaje correspondiente a la cantidad de ocurrencias de la clase dominante. Este porcentaje se utiliza cuando, dado un bosque y un ejemplo, se obtienen más de un positivo. En este caso, el criterio por el cual se decidirá a cuál árbol creerle es el de mayor porcentaje.
De la misma forma, en caso de que ningún árbol reclame la clase, se tomará a aquel cuyo porcentaje sea menor (aquel cuya probabilidad de que sea un falso negativo es mayor).

En el caso de que empaten en pocentaje, se tomara el primero.

Ejemplo:
Dada una tupla correspondiente a una _Iris Setosa_ y un bosque que contiene tres modelos de uno versus todos (uno por cada clase de _Iris_). Si al clasificar, el caso es el siguiente:

| - | Result | Percentage |
|--:|-------:|-----------:|
|Model Setosa|True|**97%**|
|Model Versicolour|True|53%|
|Model Virginica|False|96%|


Entonces la predicción (usando el criterio antes nombrado) dará como resultado _Iris Setosa_

### 3.2 Preparación del conjunto de entrenamiento

Para cada modelo de una clase versus las demas se adaptó el conjunto de entrenamiento para que el valor fuese verdadero (en caso de ser la clase que se quiere clasificar) o falso (en caso contrario).
Es decir, para generar el modelo de _Iris Setosa_ versus las demas, si en el conjunto de entrenamiento original se tenían las tuplas

```
5.0,3.3,1.4,0.2,Iris-setosa
7.0,3.2,4.7,1.4,Iris-versicolor
6.3,3.4,5.6,2.4,Iris-virginica
```
entonces, en el nuevo conjunto adaptado para generar este modelo, se tendrían las siguientes tuplas:

```
5.0,3.3,1.4,0.2,1
7.0,3.2,4.7,1.4,0
6.3,3.4,5.6,2.4,0
```

### 3.3 Evaluación

#### 3.3.1 Evaluación normal

Nuevamente, los resultados obtenidos usados para las tres métricas implementadas fueron los mismos, por lo tanto, se copian los resultados una sola vez.

Matriz de Confusión:

|-|Iris Setosa |Iris Versicolour |Iris Virginica |
|---:|---:|---:|---:|
| Iris Setosa |**8** |0 |0 |
| Iris Versicolour |2 |**3** |0 |
| Iris Virginica |0 |0 |**12** |

Valores macro y micro:

|- |Prec |Rec |Fs 0.5|
|---: |---: |---: |---:|
|Micro |0.92 |0.92 |0.92|
|Macro |0.933 |0.867 |0.899|

##### 3.3.2 Evaluación realizando 5-fold cross validation



- Presentar los distintos experimentos que se realizan y los resultados que se obtienen.

- La información de los resultados se presenta en tablas y en gráficos, de acuerdo a su naturaleza. Por ejemplo:

_En la gráfica 1, se observa el error cuadrático total del conjunto de entrenamiento a medida que pasan los juegos para el oponente X_

### 3.4 Conclusiones
Podemos observar que para los valores micro se obtuvo un 92%, mientras que para medidas macro fue del 93.3%.
Los valores indican que este metodo tuvo una buena respuesta, dado el acotado conjunto. 
Esperabamos que este metodo se destacara frente al metodo de la parte anterior, pero observando los resultados obtenididos, podemos concluir que tanto el metodo de la parte anterior, como este se comportan de forma similar.

## 4.Tercera parte 


### 4.1 2.1 Preparación del conjunto de entrenamiento
Para esta parte se tomo le conjunto _Covertype_. El cual consta de 581.012 instancias de 10 atributos cuantitativos, 2 atributos cualitativos (4 y 40 bits), y 7 distintas clases.
